packages/router/src/provide_router.ts

```diff
@@ -0,0 +1,598 @@
+ */
+
+import {LOCATION_INITIALIZED, ViewportScroller} from '@angular/common';
+import {APP_BOOTSTRAP_LISTENER, APP_INITIALIZER, ApplicationRef, ComponentRef, ENVIRONMENT_INITIALIZER, inject, InjectFlags, InjectionToken, Injector, Provider, Type} from '@angular/core';
+import {of, Subject} from 'rxjs';
+import {filter, map, take} from 'rxjs/operators';
+
+import {Event, NavigationCancel, NavigationCancellationCode, NavigationEnd, NavigationError, stringifyEvent} from './events';
+import {Routes} from './models';
+import {Router} from './router';
+import {InMemoryScrollingOptions, ROUTER_CONFIGURATION, RouterConfigOptions} from './router_config';
+import {ROUTES} from './router_config_loader';
+import {PreloadingStrategy, RouterPreloader} from './router_preloader';
+import {ROUTER_SCROLLER, RouterScroller} from './router_scroller';
+import {ActivatedRoute} from './router_state';
+
+const NG_DEV_MODE = typeof ngDevMode === 'undefined' || ngDevMode;
+
+/**
+ * Sets up providers necessary to enable `Router` functionality for the application.
+ * Allows to configure a set of routes as well as extra features that should be enabled.
+ *
+ * @usageNotes
+ *
+ * Basic example of how you can add a Router to your application:
+ * ```
+ * const appRoutes: Routes = [];
+ * bootstrapApplication(AppComponent, {
+ *   providers: [provideRouter(appRoutes)]
+ * });
+ * ```
+ *
+ * You can also enable optional features in the Router by adding functions from the `RouterFeatures`
+ * type:
+ * ```
+ * const appRoutes: Routes = [];
+ * bootstrapApplication(AppComponent,
+ *   {
+ *     providers: [
+ *       provideRouter(appRoutes,
+ *         withDebugTracing(),
+ *         withRouterConfig({paramsInheritanceStrategy: 'always'}))
+ *     ]
+ *   }
+ * );
+ * ```
+ *
+ * @see `RouterFeatures`
+ *
+ * @publicApi
+ * @developerPreview
+ * @param routes A set of `Route`s to use for the application routing table.
+ * @param features Optional features to configure additional router behaviors.
+ * @returns A set of providers to setup a Router.
+ */
+export function provideRouter(routes: Routes, ...features: RouterFeatures[]): Provider[] {
+  return [
+    provideRoutes(routes), {provide: ActivatedRoute, useFactory: rootRoute, deps: [Router]},
+    {provide: APP_BOOTSTRAP_LISTENER, multi: true, useFactory: getBootstrapListener},
+    features.map(feature => feature.ɵproviders),
+    // TODO: All options used by the `assignExtraOptionsToRouter` factory need to be reviewed for
+    // how we want them to be configured. This API doesn't currently have a way to configure them
+    // and we should decide what the _best_ way to do that is rather than just sticking with the
+    // status quo of how it's done today.
+  ];
+}
+
+export function rootRoute(router: Router): ActivatedRoute {
+  return router.routerState.root;
+}
+
+/**
+ * Helper type to represent a Router feature.
+ */
+interface RouterFeature<FeatureKind> {
```

Comment:
```suggestion
export interface RouterFeature<FeatureKind extends RouterFeatureKind> {
```

This needs to be exported as its a part of other exported types.